packages/sentinel/src/api/index.ts

```diff
@@ -55,11 +55,11 @@ export class SentinelClient extends BaseApiClient {
 
   public async update(sentinelId: string, sentinel: UpdateSentinelRequest): Promise<CreateSentinelResponse> {
     const currentSentinel = await this.get(sentinelId);
+    const updatedSentinel = await this.constructSentinelRequest(
+      _.merge(currentSentinel, sentinel) as CreateSentinelRequest,
```

Comment:
Heads up that [lodash's `merge`](https://lodash.com/docs/4.17.15#merge) is recursive! This means that nested properties can get mixed up. Also, trying to remove an item from an array would not work:

```
_.merge({ addresses: ['0x1', '0x2'] }, { addresses: ['0x1'] }
> { addresses: ['0x1', '0x2'] }
```

Here the `{ ...currentSentinel, ...sentinel }` syntax is preferred since it is not recursive.

Reply:
That's the idea 🤔 the shape of what the user put in the client is different to the shape the API receives, so if we do the spread, we're sending the `addresses` field in the first level and it's not taken when the backend parses the body and expects it to be nested in `addressRules`

Reply:
> if we do the spread, we're sending the addresses field in the first level and it's not taken when the backend parses the body and expects it to be nested in addressRules.

From what I see on the code, it's `constructBlockSentinel` who takes care of lifting `addresses` into `addressRules`. The merge itself is not related to this.

There's another thing that seems to be off: `this.get` is returning the response using the format handled by the backend (ie using `addressRules` instead of `addresses`) yet we are merging it with an object that uses the client-side format (`addresses`). This is bound to cause issues (and probably the reason why you needed the explicit cast).

If we are using a different model for representing a Sentinel on the client-side (which I think is fine, since the one on the backend is much more complicated!) we should make sure it's used _everywhere_ on the client. That means that `this.get` should be returning that model. And if `this.get` returns that simplified model, it's then safe to merge it (with a shallow merge, ie a spread) with the one that the user supplied for the update!

Reply:
> From what I see on the code, it's constructBlockSentinel who takes care of lifting addresses into addressRules. The merge itself is not related to this.

This is completely right, my bad

> There's another thing that seems to be off: this.get is returning the response using the format handled by the backend (ie using addressRules instead of addresses) yet we are merging it with an object that uses the client-side format (addresses). This is bound to cause issues (and probably the reason why you needed the explicit cast).
> If we are using a different model for representing a Sentinel on the client-side (which I think is fine, since the one on the backend is much more complicated!) we should make sure it's used everywhere on the client. That means that this.get should be returning that model. And if this.get returns that simplified model, it's then safe to merge it (with a shallow merge, ie a spread) with the one that the user supplied for the update!

Agree it's okay to have a common model we expose through the client and API but that would break backward compatibility. What I'm working on is an internal function to pass from one model to another but I lack context in a lot of fields. Not sure if @dylankilkenny may help me with this

Reply:
> Not sure if @dylankilkenny may help me with this

Yep I can help, I'll shoot you a message on slack